server/src/lib/migrate.rs

```diff
@@ -0,0 +1,326 @@
+) -> Result<(), MigrateError> {
+    info!(log, "Enter Migrate Task");
+
+    // TODO: actual migration protocol, for now just send some stuff back and forth
+    for x in 0..10 {
+        let read = conn.read_u32().await.map_err(|_| MigrateError::Protocol)?;
+        info!(log, "Src Read: {:?}", read);
+        assert_eq!(read, x);
+    }
+    for x in 10..20 {
+        conn.write_u32(x).await.map_err(|_| MigrateError::Protocol)?;
+    }
+
+    info!(log, "Migrate Successful");
+
+    Ok(())
+}
+
+/// Begin the migration process (source-side).
+///
+///This will attempt to upgrade the given HTTP request to a `propolis-migrate`
+/// connection and begin the migration in a separate task.
+pub async fn source_start(
+    rqctx: Arc<RequestContext<Context>>,
+    instance_id: Uuid,
+    migration_id: Uuid,
+) -> Result<Response<Body>, MigrateError> {
+    // Create a new log context for the migration
+    let log = rqctx.log.new(o!(
+        "migration_id" => migration_id.to_string(),
+        "migrate_role" => "source"
+    ));
+    info!(log, "Migration Source");
+
+    let mut context = rqctx.context().context.lock().await;
+    let context =
+        context.as_mut().ok_or_else(|| MigrateError::SourceNotInitialized)?;
+
+    if instance_id != context.properties.id {
+        return Err(MigrateError::UuidMismatch);
+    }
+
+    // Bail if there's already one in progress
+    // TODO: Should we just instead hold the context lock during the whole process?
+    if context.migrate_task.is_some() {
+        return Err(MigrateError::MigrationAlreadyInProgress);
+    }
+
+    let request = &mut *rqctx.request.lock().await;
+
+    // Check this is a valid migration request
+    if !request
+        .headers()
+        .get(header::CONNECTION)
+        .and_then(|hv| hv.to_str().ok())
+        .map(|hv| hv.eq_ignore_ascii_case("upgrade"))
+        .unwrap_or(false)
+    {
+        return Err(MigrateError::UpgradeExpected);
+    }
+
+    let src_protocol = MIGRATION_PROTOCOL_STR;
+    let dst_protocol = request
+        .headers()
+        .get(header::UPGRADE)
+        .ok_or_else(|| MigrateError::UpgradeExpected)
+        .map(|hv| hv.to_str().ok())?
+        .ok_or_else(|| MigrateError::incompatible(src_protocol, "<unknown>"))?;
+
+    // TODO: improve "negotiation"
+    if !dst_protocol.eq_ignore_ascii_case(MIGRATION_PROTOCOL_STR) {
+        error!(
+            log,
+            "incompatible with destination instance provided protocol ({})",
+            dst_protocol
+        );
+        return Err(MigrateError::incompatible(src_protocol, dst_protocol));
+    }
+
+    let upgrade = hyper::upgrade::on(&mut *request);
+    let instance = context.instance.clone();
+
+    // We've successfully negotiated a migration protocol w/ the destination.
+    // Now, we spawn a new task to handle the actual migration over the upgraded socket
+    let task = tokio::spawn(async move {
+        let conn = match upgrade.await {
+            Ok(upgraded) => upgraded,
```

Comment:
Could we perhaps call this `conn` or `socket` or something? I probably just need more caffeine, but I was thinking that this referred to the upgrade case for LM. :-D

Reply:
I already do have a `conn` there but I just added a comment to make it more clear. Let me know if that's enough.